ark/phenotyping/cluster_helpers.py

```diff
@@ -0,0 +1,261 @@
+import bisect
+import feather
+from itertools import combinations
+import numpy as np
+import pathlib
+import pandas as pd
+from scipy.stats import zscore
+from sklearn.cluster import AgglomerativeClustering
+from typing import Callable, List
+
+from ark.utils.misc_utils import verify_in_list
+
+###############################################
+# Copyright Žiga Sajovic, XLAB 2019           #
+# Distributed under the MIT License           #
+#                                             #
+# github.com/ZigaSajovic/Consensus_Clustering #
+#                                             #
+###############################################
+
+
+class ConsensusCluster:
+    def __init__(self, cluster: Callable, L: int, K: int, H: int,
+                 resample_proportion: float = 0.5):
+        """Implementation of Consensus clustering, following the paper
+        https://link.springer.com/content/pdf/10.1023%2FA%3A1023949509487.pdf
+
+        Args:
+            cluster (Callable):
```

Comment:
Not sure we'll need explicit error checks inside either constructor since we'll always be using `sklearn.cluster.AgglomerativeCluster`.

Is this something that can be done inside the documentation or the type hint?

Reply:
Interesting, I think what we're looking for is structural typing. I found some [documentation on it](https://peps.python.org/pep-0544/).

```python
from typing import Protocol, runtime_checkable

@runtime_checkable
class SupportsCluster(Protocol):
    def fit_predict(self) -> None: ... # Actually using the `...` 
    
    @property
    def n_clusters(self) -> int:
        return n_cluster
```

Then with typehinting you can do:
```python
class ConsensusCluster:
    def __init__(self, cluster: SupportsCluster, L: int, K: int, H: int, resample_proportion: float = 0.5):
        pass
```